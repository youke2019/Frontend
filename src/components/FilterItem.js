import React from "react";
import {
    Text,
    View,
    StyleSheet,
    Image,
    TouchableOpacity,
} from "react-native";
import {Button} from 'react-native-elements'

class FilterItem extends React.Component {
    state = {
        fold: true,
        list: this.props.list
    }

    componentDidMount() {
        for (let item of this.state.list) {
            item['selected'] = false
        }
    }

    changeFoldState = () => {
        this.setState({
            fold: !this.state.fold
        })
    }

    changeSelectState = (name) => {
        for (let item of this.state.list)
            if (item.name == name){
                item.selected = !item.selected
                this.setState({
                    list: this.state.list
                })
            }

        let filter = {}
        filter[this.props.title] = []
        for (let item of this.state.list)
            if (item.selected == true)
                filter[this.props.title].push(item.value)

        this.props.updateFilter(filter)
    }

    render() {
        const {title} = this.props
        const {list,fold} = this.state

        return (
            <View>
                <View style={styles.title_container}>
                    <Text style={{flex: 8}}>{title}</Text>
                    <TouchableOpacity
                        style={{flex:1}}
                        onPress={this.changeFoldState}>
                        <Image
                            style={styles.arrow}
                            source={{uri: fold? 'arrow_up' : 'arrow_down' }}
                        />
                    </TouchableOpacity>
                </View>
                {
                    fold &&
                    <View  style={styles.button_container}>
                        {list.map((item) =>
                            <View
                                key={item.name}
                                style={styles.button_wrap}
                            >
                                <Button
                                    title={item.name}
                                    raised={item.selected? true:false}
                                    type="clear"
                                    onPress={()=>this.changeSelectState(item.name)}
                                    buttonStyle={item.selected?styles.button_selected:styles.button_unselected}
                                    titleStyle={item.selected?styles.text_selected:styles.text_unselected}
                                />
                            </View>
                        )}
                    </View>
                }
            </View>
        );
    }
}

const styles=StyleSheet.create({
    button_container:{
        flexDirection: 'row',
        flexWrap: 'wrap',
        display:'flex',
    },
    button_wrap:{
        padding: 6
    },
    title_container:{
        flexDirection: 'row',
    },
    arrow:{
        width:30,
        height:30,
    },
    button_selected:{
        backgroundColor: '#ffe8d9',
    },
    button_unselected:{
        backgroundColor: '#F8F8FF',
    },
    text_selected:{
        color: '#ff961e',
        fontSize: 15
    },
    text_unselected:{
        color: 'black',
        fontSize: 15,
    }
})

export default FilterItem