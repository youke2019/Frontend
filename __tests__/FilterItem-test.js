import React from 'react';
import FilterItem from '../src/components/FilterItem';

import renderer from 'react-test-renderer';
import Adapter from 'enzyme-adapter-react-16'
import {shallow, mount, render, configure} from 'enzyme'

configure({adapter: new Adapter()})

const updateFilter = jest.fn((filter) => {

})

const list = {
    credits:[
        {name:'1',value:'1'},
        {name:'2',value:'2'},
        {name:'3',value:'3'},
        {name:'4',value:'4'},
    ]
}

test('renders FilterItem correctly', () => {
    const tree = renderer.create(<FilterItem
        title="学分"
        list={list.credits}
        updateFilter={(filter) => updateFilter(filter)}
    />).toJSON();
    expect(tree).toMatchSnapshot();
});

test('test function changeFoldState()', () => {
    let wrapper = shallow(<FilterItem
    title="学分"
    list={list.credits}
    updateFilter={(filter) => updateFilter(filter)}
    />)

    wrapper.find('TouchableOpacity').prop('onPress')()
    expect(wrapper.state().fold).toBe(false)
});

test('test function changeSelectState()', () => {
    let wrapper = shallow(<FilterItem
        title="学分"
        list={list.credits}
        updateFilter={(filter) => updateFilter(filter)}
    />)

    const button = { name: '1', value: '1', selected: true }
    wrapper.find({title:'1'}).props().onPress()
    expect(wrapper.state().list).toContainEqual(button)
    expect(updateFilter).toHaveBeenCalled()
});