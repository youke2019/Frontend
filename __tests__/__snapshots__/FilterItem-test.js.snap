// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders FilterItem correctly 1`] = `
<View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "padding": 8,
      }
    }
  >
    <Text
      style={
        Object {
          "flex": 8,
          "fontSize": 16,
        }
      }
    >
      学分
    </Text>
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "flex": 1,
          "opacity": 1,
        }
      }
    >
      <Image
        source={
          Object {
            "uri": "arrow_down",
          }
        }
        style={
          Object {
            "height": 21,
            "width": 21,
          }
        }
      />
    </View>
  </View>
</View>
`;
